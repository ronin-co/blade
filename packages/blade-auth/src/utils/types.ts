import type { Account } from 'blade/types';

type AppendArg<F, A> = F extends (this: infer T, ...args: infer P) => infer R
  ? (this: T, ...args: [...P, A]) => R
  : never;

type EmailType = 'PASSWORD_RESET' | 'EMAIL_VERIFICATION';
type EmailOptions = { account: Account; type: EmailType; token: string };

interface AuthOptions {
  sendEmail?: (options: EmailOptions) => Promise<void>;
}

export type WithAuthOptions<Trigger> = AppendArg<Trigger, AuthOptions>;
