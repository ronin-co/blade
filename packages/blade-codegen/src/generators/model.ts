import { SyntaxKind, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { triggerOptionsInterface } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type { TypeAliasDeclaration, TypeParameterDeclaration } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { PopulatedModel } from 'blade-compiler';

/**
 * Generate all required type definitions for a provided schema model.
 *
 * The plural model type will be mapped to an array of the singular model type and
 * extended with the plural model properties.
 *
 * @param models - All models of the addressed space.
 *
 * @returns An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateModelTypes = (
  models: Array<PopulatedModel>,
): Array<TypeAliasDeclaration> => {
  const nodes = new Array<TypeAliasDeclaration>(triggerOptionsInterface);

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const modelIdentifier = {
      singular: factory.createIdentifier(convertToPascalCase(model.slug)),
      plural: factory.createIdentifier(convertToPascalCase(model.pluralSlug)),
    };

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) =>
        factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
      );

    /**
     * ```ts
     * <U extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      typeArgumentIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<U>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      modelIdentifier.singular,
      hasLinkFields
        ? [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)]
        : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<U extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.singular,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createTypeReferenceNode(
          factory.createQualifiedName(
            identifiers.namespace.utils.name,
            identifiers.namespace.utils.resultRecord,
          ),
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<U extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.plural,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    nodes.push(singularModelTypeDec, pluralModelTypeDec);
  }

  return nodes;
};
