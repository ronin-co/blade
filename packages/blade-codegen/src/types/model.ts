import type { ModelField as PartialModelField } from 'blade-compiler';

type RecursiveRequired<T> = {
  [K in keyof T]-?: T[K] extends object ? RecursiveRequired<T[K]> : T[K];
};

export type ModelField = RecursiveRequired<PartialModelField>;
