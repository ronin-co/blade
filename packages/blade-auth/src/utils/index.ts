import { EmptyFieldsError, InvalidPermissionsError } from 'blade/server/errors';
import { type JWTPayload, verifyJWT } from 'blade/server/utils';
import type { SetQueryInstructions, TriggerOptions } from 'blade/types';

/**
 * Check given object for empty values. Throws an error if the object contains
 * empty values.
 *
 * @param to Object to check for empty values.
 * @param allowed Optional array of keys to exclude. Used in cases where `null`
 * is allowed as a value.
 */
export const avoidEmptyFields = (
  to: SetQueryInstructions['to'],
  allowed?: Array<string>,
) => {
  // We'll no longer have to use `(to || {})` once the `to` property on
  // `SetQueryInstructions` is guaranteed to be defined, which it already
  // should be.
  const emptyFields = Object.keys(to || {}).filter(
    (key) =>
      !(allowed || []).includes(key) && typeof to?.[key] !== 'boolean' && !to?.[key],
  );

  if (emptyFields.length > 0) {
    throw new EmptyFieldsError({ fields: emptyFields });
  }
};

/**
 * Obtains the secret used to sign authentication tokens.
 *
 * @returns The secret.
 */
export const getAuthSecret = (): string => {
  let secret = import.meta.env.BLADE_AUTH_SECRET;

  if (!secret) {
    if (import.meta.env.BLADE_ENV === 'development') {
      secret = 'default-secret-1234';
    } else {
      throw new Error('Please add a `BLADE_AUTH_SECRET` environment variable.');
    }
  }

  return secret;
};

/**
 * Retrieve the account that authored the incoming query.
 *
 * @param triggerOptions - The options object passed to the surrounding effect.
 *
 * @returns The ID of the account that authored the incoming query.
 */
export const getSessionCookie = async (
  triggerOptions: TriggerOptions,
): Promise<JWTPayload> => {
  const token = triggerOptions.cookies.session;

  let sessionId: string | null = null;
  let accountId: string | null = null;

  if (token) {
    try {
      const tokenPayload = await verifyJWT(token, getAuthSecret());

      sessionId = (tokenPayload?.sub as string) || null;
      accountId = (tokenPayload?.aud as string) || null;
    } catch (err) {
      console.warn(`Failed to decode token: \`${(err as Error).message}\``);
    }
  }

  if (!(sessionId && accountId)) {
    throw new InvalidPermissionsError({
      message: 'No session provided for authentication.',
    });
  }

  return { sessionId, accountId };
};

/**
 * Generate pseudo-random unique identifiers.
 *
 * @returns A unique identifier.
 */
export const generateUniqueId = (): string => {
  return crypto.getRandomValues(new Uint32Array(1))[0].toString();
};

export const EMAIL_VERIFICATION_COOLDOWN = 20_000; // 20s
