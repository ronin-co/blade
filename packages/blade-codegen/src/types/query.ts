import type { DML_QUERY_TYPES } from 'blade-compiler';

export type QueryType = (typeof DML_QUERY_TYPES)[number] | 'use' | 'useMutation';
