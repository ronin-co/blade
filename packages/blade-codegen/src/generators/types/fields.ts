import { factory } from 'typescript';

import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { convertToPascalCase } from '@/src/utils/slug';

import type { TypeAliasDeclaration } from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Generates TypeScript type aliases for model field slugs.
 *
 * @example
 * ```ts
 * type UserFieldSlug = 'id' | 'ronin.createdAt' | [...] | 'name' | 'email';
 * ```
 *
 * @param models - The models for which to generate field slug types.
 *
 * @returns An array of TypeScript type alias declarations for each model's field slugs.
 */
export const generateModelFieldsTypes = (
  models: Array<Model>,
): Array<TypeAliasDeclaration> =>
  models.map((model) =>
    factory.createTypeAliasDeclaration(
      undefined,
      factory.createIdentifier(`${convertToPascalCase(model.slug)}FieldSlug`),
      undefined,
      factory.createUnionTypeNode(
        [...DEFAULT_FIELD_SLUGS, ...Object.keys(model.fields)].map((slug) =>
          factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
        ),
      ),
    ),
  );
