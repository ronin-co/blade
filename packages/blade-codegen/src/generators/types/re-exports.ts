import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';
import { convertToPascalCase } from '@/src/utils/slug';

import type { TypeAliasDeclaration } from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Generate type re-exports for all provided models.
 *
 * @example
 * ```ts
 * export type User = import('blade/types').User;
 * export type Users = import('blade/types').Users;
 * ```
 *
 * @param model - The model to generate type re-exports for.
 *
 * @returns An array of type alias declarations for each model.
 */
export const generateTypeReExports = (
  model: Model,
): [TypeAliasDeclaration, TypeAliasDeclaration] => {
  const slug = convertToPascalCase(model.slug);
  const pluralSlug = convertToPascalCase(model.pluralSlug);

  return [
    factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      slug,
      undefined,
      factory.createImportTypeNode(
        factory.createTypeReferenceNode(identifiers.blade.module.types),
        undefined,
        factory.createIdentifier(slug),
        undefined,
        false,
      ),
    ),
    factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      pluralSlug,
      undefined,
      factory.createImportTypeNode(
        factory.createTypeReferenceNode(identifiers.blade.module.types),
        undefined,
        factory.createIdentifier(pluralSlug),
        undefined,
        false,
      ),
    ),
  ];
};
