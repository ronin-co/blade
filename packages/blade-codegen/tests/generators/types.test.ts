import { describe, expect, test } from 'bun:test';
import {
  blob,
  boolean,
  date,
  json,
  link,
  model,
  number,
  string,
} from 'blade-syntax/schema';
import { factory } from 'typescript';

import { generateModelTypes } from '@/src/generators/types/model';
import {
  generateDefaultSyntaxProperty,
  generateOrderedBySyntaxProperty,
  generateRootQueryCallSignature,
  generateSelectingSyntaxProperty,
  generateUsingSyntaxProperty,
  generateWithSyntaxProperty,
} from '@/src/generators/types/syntax';
import { printNodes } from '@/src/utils/print';

import type { Model } from '@/src/types/model';

describe('types', () => {
  describe('generate the core model types', () => {
    test('a basic model', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          avatar: blob(),
          email: string({ required: true }),
          isActive: boolean(),
          lastActiveAt: date(),
          name: string(),
          rewardPoints: number({ defaultValue: 0, required: true }),
          settings: json({ defaultValue: {}, required: true }),
        },
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel], AccountModel);

      expect(typesResult).toHaveLength(2);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with a summary', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
        },
        // @ts-expect-error This property is not native to RONIN models.
        summary: 'A user account.',
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel], AccountModel);

      expect(typesResult).toHaveLength(2);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with an invalid field type', () => {
      const field = string();

      // @ts-expect-error These properties are designed to be read-only.
      field.type = 'invalid';

      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: field,
        },
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel], AccountModel);

      expect(typesResult).toHaveLength(2);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with a link field', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
        },
      });

      const PostModel = model({
        slug: 'post',
        pluralSlug: 'posts',
        fields: {
          title: string(),
          author: link({ target: 'account' }),
        },
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel, PostModel], PostModel);

      expect(typesResult).toHaveLength(4);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with a many-to-many link field', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
        },
      });

      const SpaceModel = model({
        slug: 'space',
        pluralSlug: 'spaces',
        fields: {
          name: string(),
          members: link({ target: 'account', kind: 'many' }),
        },
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel, SpaceModel], SpaceModel);

      expect(typesResult).toHaveLength(4);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with a link field that does not exist', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
          latestPost: link({ target: 'does_not_exist' }),
        },
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel], AccountModel);

      expect(typesResult).toHaveLength(2);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });

    test('a model with nested fields', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          'nested.foo': string(),
          'nested.bar': number(),
        },
        // @ts-expect-error This property is not native to RONIN models.
        summary: 'A user account.',
      });

      // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
      // @ts-expect-error Codegen models types differ from the schema model types.
      const typesResult = generateModelTypes([AccountModel], AccountModel);

      expect(typesResult).toHaveLength(2);

      const typesResultStr = printNodes(typesResult);

      expect(typesResultStr).toMatchSnapshot();
    });
  });

  describe('syntax', () => {
    describe('generate root query call signature', () => {
      test('a basic model', () => {
        const typesResult = generateRootQueryCallSignature(
          factory.createTypeReferenceNode('Account'),
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
    describe('generate `default` syntax property', () => {
      test('a basic model', () => {
        const typesResult = generateDefaultSyntaxProperty(
          'including',
          factory.createTypeReferenceNode('Account'),
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
    describe('generate `orderedBy` syntax property', () => {
      test('a basic model', () => {
        const AccountModel = model({
          slug: 'account',
          pluralSlug: 'accounts',
          fields: {
            avatar: blob(),
            email: string({ required: true }),
            isActive: boolean(),
            lastActiveAt: date(),
            name: string(),
            rewardPoints: number({ defaultValue: 0, required: true }),
            settings: json({ defaultValue: {}, required: true }),
          },
        }) as unknown as Model;

        const typesResult = generateOrderedBySyntaxProperty(
          AccountModel,
          factory.createTypeReferenceNode(AccountModel.slug),
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
    describe('generate `selecting` syntax property', () => {
      test('a basic model', () => {
        const AccountModel = model({
          slug: 'account',
          pluralSlug: 'accounts',
          fields: {
            avatar: blob(),
            email: string({ required: true }),
            isActive: boolean(),
            lastActiveAt: date(),
            name: string(),
            rewardPoints: number({ defaultValue: 0, required: true }),
            settings: json({ defaultValue: {}, required: true }),
          },
        }) as unknown as Model;

        const typesResult = generateSelectingSyntaxProperty(
          AccountModel,
          factory.createTypeReferenceNode(AccountModel.slug),
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
    describe('generate `using` syntax property', () => {
      test('a basic model', () => {
        const PostModel = model({
          slug: 'post',
          pluralSlug: 'posts',
          fields: {
            title: string(),
            author: link({ target: 'account' }),
          },
        }) as unknown as Model;

        const singularTypesResult = generateUsingSyntaxProperty(
          PostModel,
          factory.createTypeReferenceNode('Post'),
          PostModel.slug,
          false,
        );
        const pluralTypesResult = generateUsingSyntaxProperty(
          PostModel,
          factory.createTypeReferenceNode('Post'),
          PostModel.slug,
          true,
        );

        const typesResultStr = printNodes([singularTypesResult, pluralTypesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });

      test('a basic model with no link fields', () => {
        const AccountModel = model({
          slug: 'account',
          pluralSlug: 'accounts',
          fields: {
            avatar: blob(),
            email: string({ required: true }),
            isActive: boolean(),
            lastActiveAt: date(),
            name: string(),
            rewardPoints: number({ defaultValue: 0, required: true }),
            settings: json({ defaultValue: {}, required: true }),
          },
        }) as unknown as Model;

        const singularTypesResult = generateUsingSyntaxProperty(
          AccountModel,
          factory.createTypeReferenceNode('Account'),
          AccountModel.slug,
          false,
        );
        const pluralTypesResult = generateUsingSyntaxProperty(
          AccountModel,
          factory.createTypeReferenceNode('Account'),
          AccountModel.slug,
          true,
        );

        const typesResultStr = printNodes([singularTypesResult, pluralTypesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
    describe('generate `with` syntax property', () => {
      test('a basic model', () => {
        const AccountModel = model({
          slug: 'account',
          pluralSlug: 'accounts',
          fields: {
            avatar: blob(),
            email: string({ required: true }),
            isActive: boolean(),
            lastActiveAt: date(),
            name: string(),
            rewardPoints: number({ defaultValue: 0, required: true }),
            settings: json({ defaultValue: {}, required: true }),
          },
        }) as unknown as Model;

        const typesResult = generateWithSyntaxProperty(
          AccountModel,
          factory.createTypeReferenceNode('Account'),
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
  });
});
