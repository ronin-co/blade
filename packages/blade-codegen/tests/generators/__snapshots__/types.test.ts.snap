// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`types generate the core model types a basic model 1`] = `
"export type Account = ResultRecord & {
    avatar: StoredObject;
    email: string;
    isActive: boolean;
    lastActiveAt: Date;
    name: string;
    rewardPoints: number;
    settings: (JsonObject | JsonArray);
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with a summary 1`] = `
"/**
 * A user account.
 */
export type Account = ResultRecord & {
    email: string;
    name: string;
};
/**
 * A user account.
 */
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with an invalid field type 1`] = `
"export type Account = ResultRecord & {
    name: unknown;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with a link field 1`] = `
"export type Account = ResultRecord & {
    email: string;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
export type Post<U extends Array<"author"> | "all" = [
]> = ResultRecord & {
    author: ResolveSchema<Account, U, "author">;
    title: string;
};
export type Posts<U extends Array<"author"> | "all" = [
]> = Array<Post<U>> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with a many-to-many link field 1`] = `
"export type Account = ResultRecord & {
    email: string;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
export type Space<U extends Array<"members"> | "all" = [
]> = ResultRecord & {
    members: ResolveSchema<Array<Account>, U, "members">;
    name: string;
};
export type Spaces<U extends Array<"members"> | "all" = [
]> = Array<Space<U>> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with a link field that does not exist 1`] = `
"export type Account = ResultRecord & {
    email: string;
    latestPost: unknown;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types generate the core model types a model with nested fields 1`] = `
"/**
 * A user account.
 */
export type Account = ResultRecord & {
    name: string;
    nested: {
        bar: number;
        foo: string;
    };
};
/**
 * A user account.
 */
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types syntax generate root query call signature a basic model 1`] = `
"<T = Account>(instructions?: Partial<CombinedInstructions>, options?: Record<string, unknown>): T;
"
`;

exports[`types syntax generate \`default\` syntax property a basic model 1`] = `
"including: ReducedFunction & (<T = Account>(value: CombinedInstructions["including"]) => T);
"
`;

exports[`types syntax generate \`orderedBy\` syntax property a basic model 1`] = `
"orderedBy: ReducedFunction & (<T = account>(options: {
    ascending?: Array<Expression | AccountFieldSlug>;
    descending?: Array<Expression | AccountFieldSlug>;
}) => T) & {
    ascending: <T = account>(fields: Array<Expression | AccountFieldSlug>) => T;
    descending: <T = account>(fields: Array<Expression | AccountFieldSlug>) => T;
};
"
`;

exports[`types syntax generate \`selecting\` syntax property a basic model 1`] = `
"selecting: ReducedFunction & (<T = account>(options: Array<AccountFieldSlug>) => T);
"
`;

exports[`types syntax generate \`using\` syntax property a basic model 1`] = `
"using: ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): post<U> | null;
    <T = Post>(fields: Array<"author"> | "all"): T;
};
using: ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): Promise<post<U>>;
    <T = Post>(fields: Array<"author"> | "all"): Promise<T>;
};
"
`;

exports[`types syntax generate \`using\` syntax property a basic model with no link fields 1`] = `
"using: ReducedFunction & (<T = Account>(value: CombinedInstructions["using"]) => T);
using: ReducedFunction & (<T = Account>(value: CombinedInstructions["using"]) => T);
"
`;

exports[`types syntax generate \`with\` syntax property a basic model 1`] = `
"with: ReducedFunction & {
    <T = Account>(instructions: CombinedInstructions["with"], options?: Record<string, unknown>): T;
    id: <T = Account>(value: ResultRecord["id"]) => T;
    ronin: ReducedFunction & {
        createdAt: <T = Account>(value: ResultRecord["ronin"]["createdAt"]) => T;
        createdBy: <T = Account>(value: ResultRecord["ronin"]["createdBy"]) => T;
        updatedAt: <T = Account>(value: ResultRecord["ronin"]["updatedAt"]) => T;
        updatedBy: <T = Account>(value: ResultRecord["ronin"]["updatedBy"]) => T;
    };
    avatar: <T = Account>(avatar: Account["avatar"]) => T;
    email: <T = Account>(email: Account["email"]) => T;
    isActive: <T = Account>(isActive: Account["isActive"]) => T;
    lastActiveAt: <T = Account>(lastActiveAt: Account["lastActiveAt"]) => T;
    name: <T = Account>(name: Account["name"]) => T;
    rewardPoints: <T = Account>(rewardPoints: Account["rewardPoints"]) => T;
    settings: <T = Account>(settings: Account["settings"]) => T;
};
"
`;
